pydatastructs/linear_data_structures/algorithms.py

```diff
@@ -233,3 +234,72 @@ def brick_sort_parallel(array, num_threads, **kwargs):
 
     if _check_type(array, DynamicArray):
         array._modify(force=True)
+
+def _matrix_multiply_helper(m1, m2, row, col):
+    s = 0
+    for i in range(len(m1)):
+        s += m1[row][i] * m2[i][col]
+    return s
+
+def matrix_multiply_parallel(matrix_1, matrix_2, num_threads=None):
+    """
+    Implementes concurrent Matrix multiplication
+
+    Parameters
+    ==========
+
+    matrix_1: list of list / tuple of tuple / OneDimensionalArray of OneDimensionalArray
+        Left matrix
+
+    matrix_1: list of list / tuple of tuple / OneDimensionalArray of OneDimensionalArray
+        Right matrix
+
+    num_threads: int
+        The maximum number of threads
+        to be used for multiplication.
+        If not specified then assumed to be equal to number of rows.
+
+    Raises
+    ======
+
+    IndexError
+        When the columns in matrix_1 are not equal to the rows in matrix_2
+
+    Returns
+    =======
+
+    Matrix: Resultant matrix of matrix_1 * matrix_2
+
+    Examples
+    ========
+
+    >>> from pydatastructs import matrix_multiply_parallel
+    >>> I = [[1, 1, 0], [0, 1, 0], [0, 0, 1]]
+    >>> J = [[2, 1, 2], [1, 2, 1], [2, 2, 2]]
+    >>> print(matrix_multiply_parallel(I, J))
+    [[3, 3, 3], [1, 2, 1], [2, 2, 2]]
+
+    References
+    ==========
+    .. [1] https://en.wikipedia.org/wiki/Matrix_multiplication_algorithm#Parallel_and_distributed_algorithms
+    """
+    row_matrix_1, col_matrix_1 = len(matrix_1), len(matrix_1[0])
+    row_matrix_2, col_matrix_2 = len(matrix_2), len(matrix_2[0])
+
+    if col_matrix_1 != row_matrix_2:
+        raise IndexError("Matrices cannot be multiplied")
+
+    C = [[None for i in range(col_matrix_1)] for j in range(row_matrix_2)]
+
+    if(not num_threads):
+        num_threads = row_matrix_1
+
+    with ThreadPoolExecutor(max_workers=num_threads) as Executor:
+        for i in range(row_matrix_1):
+            for j in range(col_matrix_2):
+                C[i][j] = Executor.submit(_matrix_multiply_helper,
+                                          matrix_1,
+                                          matrix_2,
+                                          i, j).result()
```

Comment:
```suggestion
                C[i].append(Executor.submit(_matrix_multiply_helper,
                                          matrix_1,
                                          matrix_2,
                                          i, j).result())
```

Reply:
It is asymptotically O(1) but initializing it with None is exactly O(1). this method would be slower for small matrices

Reply:
> It is asymptotically O(1)

IMHO, there is nothing like exactly O(1), O itself is asymptotic. In addition, `append` takes amortized O(1) time and the logic in the suggestion may lead to inconsistencies. Resolving.